### PR TITLE
feat(webpack-builder): basic e2e solution

### PR DESCRIPTION
**Pull request by Asuka109**
# PR Details

## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] I have updated changeset
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: packages/builder/webpack-builder/src/shared/fs.ts [hunk 1]
```diff
@@ -76,3 +78,30 @@ export const getFilename = (
       throw new Error(`unknown key ${type} in "output.filename"`);
   }
 };
+
+export interface GlobContentJSONOptions extends GlobbyOptions {
+  maxSize?: number;
+}
+
+export const globContentJSON = async (
```
[the remark this answers is not in the document]
Comment (Asuka109) [reply packages/builder/webpack-builder/src/shared/fs.ts]: 移到 src/stub/utils.ts 了

File: packages/builder/webpack-builder/src/shared/constants.ts [hunk 2]
```diff
@@ -36,3 +36,13 @@ export const MEDIA_EXTENSIONS = [
 ];
 
 export const RUNTIME_CHUNK_NAME = 'builder-runtime';
+
+export const STUB_BUILDER_PLUGIN_BUILTIN =
```
[the remark this answers is not in the document]
Comment (Asuka109) [reply packages/builder/webpack-builder/src/shared/constants.ts]: 移到 src/stub/constants.ts 了

File: packages/builder/webpack-builder/package.json [hunk 3]
```diff
@@ -96,8 +103,18 @@
     "@types/serialize-javascript": "^5.0.1",
     "@vitest/ui": "^0.21.1",
     "autoprefixer": "^10.4.8",
+    "memfs": "^3.3.0",
     "typescript": "^4",
-    "vitest": "^0.21.1"
+    "vitest": "^0.21.1",
+    "@modern-js/e2e": "workspace:*"
+  },
+  "peerDependencies": {
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/package.json:111]: 为什么 builder 需要在 peerDependencies 里声明 e2e 嘞

File: packages/builder/webpack-builder/src/shared/fs.ts [hunk 4]
```diff
@@ -76,3 +77,6 @@ export const getFilename = (
       throw new Error(`unknown key ${type} in "output.filename"`);
   }
 };
+
+export const filenameToGlobExpr = (file: PathLike) =>
```
Comment (chenjiahan) [inline packages/builder/webpack-builder/src/shared/fs.ts:81]: 这个也是 stub 用的函数喔，建议放 stub 目录下

File: tests/e2e/webpack-builder/basic/package.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+  "private": "true",
```
[the remark this answers is not in the document]
Comment (chenjiahan) [reply tests/e2e/webpack-builder/basic/package.json:2]: 这个地方是不需要修改嘛

File: packages/builder/webpack-builder/package.json (hunk 3 above)
Comment (Asuka109) [reply packages/builder/webpack-builder/package.json:111]: 设想是把 e2e 相关能力隔离到 e2e 包里，e2e 作为对用户透出能力的包，同时也是 builder 的可选依赖用于共享代码。目前主要是 https://github.com/modern-js-dev/modern.js/blob/2d02999c8603157456e1095589482ded6c2211b0/packages/builder/webpack-builder/src/stub/builder.ts#L228 这样的用法。

Comment (Asuka109) [reply packages/builder/webpack-builder/package.json:111]: https://github.com/modern-js-dev/modern.js/blob/9d4caac63dcf7af58d99e1ca9787460ee3ad601c/packages/builder/webpack-builder/package.json#L111-L118

这样设置之后预期任何包管理器都不会自动安装 e2e 包，而是让用户在需要时手动安装

File: packages/builder/webpack-builder/src/shared/fs.ts (hunk 4 above)
Comment (Asuka109) [reply packages/builder/webpack-builder/src/shared/fs.ts:81]: 🆗

File: tests/e2e/webpack-builder/basic/package.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (Asuka109) [reply tests/e2e/webpack-builder/basic/package.json]: @chenjiahan 我改了别处漏掉这了，看了下好几个包都是这样，我一块儿改了吧 😂
image

**Review by chenjiahan — APPROVED** [review]

**Comment by chenjiahan** [comment]
👏🏻👏🏻

**Pull request merged**